### DISCUSSION TITLE
Are you still updating this?

### DISCUSSION THREAD

**Discussion by alecmce** [protectwise/troika#146] [General]
Hi, I heard that the main contributor to this is no longer part of protectwise. I have a modest API change that I'd like to propose. Is it worth pursuing?

**Comment by lojjic** [comment]
@alecmce You heard correctly, but I will continue contributing. What's your API change proposal?

**Reply by alecmce** [reply]
The `sync` method receives a callback but you can't guarantee that the callback will ever be called. So, external to `sync` we need to check whether `_needsSync`, and call the callback if it does not. Minimally, I'd like not to have to access a value that's "private adjacent" ... In our code that consumes `troika-three-text`, it looks for all the world like a bug that we're accessing a value with an underscore. So, my minimal change would be to expose a `getNeedsSync()` function that returns `_needsSync`. I mean, I'd really like to guarantee that the callback is called when `sync` is called, but I understand that's an API-breaking change.

**Reply by lojjic** [reply]
Is your use case for the callback something that wouldn't be served by the [synccomplete event listener](https://protectwise.github.io/troika/troika-three-text/#handling-asynchronous-updates)? I usually recommend that as a more robust solution because it is guaranteed to fire whenever the rendering is updated.

**Reply by alecmce** [reply]
The synccomplete event only gets called when `_needsSync`, though? The use-case of "I don't know whether you need to sync or not troika, but if you do, sync, and when you're done, trigger this callback ... if not, just trigger this callback!" is what I really want. A "promise", more or less. To do that I can't see any way around reading `_needsSync` directly. But the `_` implies I shouldn't be doing that.

**Reply by lojjic** [reply]
OK thanks, I think I can envision your use case now. I've only ever needed to react to changes and not cared about the no-changes case, but I can see potential uses where you'd want to wait for the state to be complete before continuing with your logic.

I'd be fine with:
- adding a public `needsSync` property you can check
- making the `sync()` method return a Promise, which would be resolved when the text is next in "synced" state (right away if it doesn't need a sync.)
- ... both?
- ... something else?

**Reply by lojjic** [reply]
Actually rather than a `needsSync` boolean, we'd probably want a new `status` enum. Something like `{needs_sync, syncing, synced}` to handle the intermediate case.

**Reply by alecmce** [reply]
I've ended up porting the main Text.js api to TypeScript within my our project and refactoring the sync mechanism for our own purposes, so I'm not looking for a change in this project anymore. For what it's worth, I'm broadly doing:

```ts
sync(callback: ApplyTextRenderInfo = NULL_FN): void {
  if (this.isValid) {
    callback(this.info)
  } else {
    this.queue.push(callback)
    this.debounce(this.flush)
  }
}

flush = (): void => {
  const queue = [...this.queue]
  this.queue.length = 0

  getTextRenderInfo(this.state, info => {
    queue.forEach(fn => fn(this.info))
  })
}
```

though there's a problem with this approach that if you sync1/invalidate/sync2/response1/invalidate/sync3, the sync3 callback will be triggered with the response1, so I need to fix that race-condition. I'll check with my boss whether it's ok to share the actual code later. The debounce is because different modules of code may have a reference to a single Text instance and may manipulate different aspects of it, then perform different actions based on the text render info that is returned.

Anyway, thank you for the very effective implementation. The API doesn't quite fit our use-case, but in the end it's probably simpler that we rewrite the API while standing on the shoulders of your implementation, rather than seek to modify your API.
